pkg/storage/wal/segment.go

```diff
@@ -49,35 +57,39 @@ type streamSegment struct {
 }
 
 // NewWalSegmentWriter creates a new WalSegmentWriter.
-func NewWalSegmentWriter() *SegmentWriter {
-	return &SegmentWriter{
-		streams: swiss.NewMap[streamID, *streamSegment](64),
-		buf1:    encoding.EncWith(make([]byte, 0, 4)),
+func NewWalSegmentWriter() (*SegmentWriter, error) {
+	idxWriter, err := index.NewWriter(context.TODO())
+	if err != nil {
+		return nil, err
 	}
+	return &SegmentWriter{
+		streams:   make(map[streamID]*streamSegment, 64),
+		buf1:      encoding.EncWith(make([]byte, 0, 4)),
+		idxWriter: idxWriter,
+	}, nil
 }
 
 // Labels are passed a string  `{foo="bar",baz="qux"}`  `{foo="foo",baz="foo"}`. labels.Labels => Symbols foo, baz , qux
 func (b *SegmentWriter) Append(tenantID, labelsString string, lbls labels.Labels, entries []*logproto.Entry) {
 	if len(entries) == 0 {
 		return
 	}
+	for _, e := range entries {
+		b.inputSize += int64(len(e.Line))
+	}
 	id := streamID{labels: labelsString, tenant: tenantID}
-	s, ok := b.streams.Get(id)
+	s, ok := b.streams[id]
 	if !ok {
 		if lbls.Get(tsdb.TenantLabel) == "" {
 			lbls = labels.NewBuilder(lbls).Set(tsdb.TenantLabel, tenantID).Labels()
 		}
-		s = &streamSegment{
-			// todo: should be pooled.
-			// prometheus bucketed pool
-			// https://pkg.go.dev/github.com/prometheus/prometheus/util/pool
-			entries:  make([]*logproto.Entry, 0, 64),
-			lbls:     lbls,
-			tenantID: tenantID,
-		}
+		s = streamSegmentPool.Get().(*streamSegment)
+		s.lbls = lbls
+		s.tenantID = tenantID
+		s.entries = s.entries[:0]
```

Comment:
Should a streamSegment have a `reset()` method?
It looks like only one line to reset them but it would maintain consistency with our other Readers/Writers.